api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,93 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+// A MatchTree defines a matching tree that can be used to match incoming data, arriving at a match action that indicates what action to take in response to the match result.
+// This structure allows for traversal to happen as data is becoming available, allowing for optimized matching.
+//
+// We talk about an "associated operation" throughout this docs: this is the context dependent operation that is configured with this match tree. For example, this could be a HTTP/TCP filter.
+message MatchTree {
+  // The action that should be taken upon matching.
+  message MatchAction {
+    oneof action {
+      option (validate.required) = true;
+
+      // The associated operation should be skipped.
+      bool skip = 1 [(validate.rules).bool = {const: true}];
+
+      // The specified callback value should be passed to the associated operation.
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    message LeafMatcher {
+      oneof matcher {
+        option (validate.required) = true;
+
+        MatchPredicate predicate = 1;
```

Comment:
Open question: Should we reuse this proto? It currently has HTTP awareness, but is set up for further flexibility so we could add a generic multimap matcher to it as well. 

Alternatively we could create a net new proto that looks the same but that doesn't carry with it the same API debt

Reply:
Sorry what is the API debt you are referring to? I think we should reuse the existing proto if we can. If we can't that's fine but it would be nice to better understand why. I think I still have unresolved questions about whether this existing proto is going to work OK since it combines HTTP request and response in the same message (without some type of capability control to make sure we don't start buffering unless the user explicitly asks for it).

Also, can you make sure every message and field has documentation? It will make it easier to read/follow.

Reply:
You're going to need to switch to whatever is being done in https://github.com/envoyproxy/envoy/pull/12938. This is part of the API tech debt, e.g. it move to StringMatchers.

Also, do we want to hardcode HTTP here? If so, why?

Reply:
Today `MatchPredicate` contains
```
    // HTTP request headers match configuration.
    HttpHeadersMatch http_request_headers_match = 5;

    // HTTP request trailers match configuration.
    HttpHeadersMatch http_request_trailers_match = 6;

    // HTTP response headers match configuration.
    HttpHeadersMatch http_response_headers_match = 7;

    // HTTP response trailers match configuration.
    HttpHeadersMatch http_response_trailers_match = 8;

    // HTTP request generic body match configuration.
    HttpGenericBodyMatch http_request_generic_body_match = 9;

    // HTTP response generic body match configuration.
    HttpGenericBodyMatch http_response_generic_body_match = 10
```

which supports HTTP in a non-generic way. If we then want to extend this to support matching on arbitrary multimaps similar to how we're supporting it via the sublinear matcher, then we'd most likely add another oneof entry with `GenericMultimapMatcher` and deprecate all the HTTP ones. This leaves us with two ways of configuring HTTP matches for this proto, potentially going through deprecation for one of them. If we start fresh and never support these HTTP specific matchers we avoid that accumulation of debt.

>I think I still have unresolved questions about whether this existing proto is going to work OK since it combines HTTP request and response in the same message (without some type of capability control to make sure we don't start buffering unless the user explicitly asks for it).

How would the user ask for this in any other way than specifying a match criteria that requires matching on buffered data? In my mind we could just look at what they configure for the match tree and if we need to buffer to determine the result we'll buffer, in which case having it all in a single proto should be fine? I might be missing something.

Reply:
I think by not special casing HTTP, it will force us to work through the quirks necessary to provide a generic solution that can be expressive enough for an HTTP-like protocol. I think it would be fine to probably do something like `MatchPredicate` but `s/header/key`, i.e. it needs to be something functionally equivalent but just drop the HTTP parlance. I wonder if the `MatchPredicate` fork in https://github.com/envoyproxy/envoy/pull/12938 is the right place to start introducing something like this.

Reply:
Another thing on this topic, it looks like we actually have two qualitatively different ways to match on headers (or keys). At leaves, we have something that is `MatchPredicate`-like. At inner nodes, we have a more limited `MultiMapMatcher`. Is there a good reason for this distinction? Is this because it's that much more expensive to do the more general `MatchPredicate` in a structured way for inner nodes?

I worry that this mix of expressiveness will cause folks to do gymnastics. Maybe we should have a short doc listing example match structures and how they get turned into this `MatchTree`?

Reply:
The distinction in my mind is about whether you're trying to get a subtree or a result out of the matching. The inner node matcher defines the branching logic while the leaf nodes define a more robust match criteria that results in the final action.

I think it would be feasible to have all matchers (both leaf and node) to both be able to return either a subtree or a final result, and support both the full matching via the predicate or the more efficient matching via the constant time lookups as part of a oneof. This would remove the distinction between leaf and node matchers, which simplifies the API in some ways, though the lack of structure might make it less obvious how to use it and provide fewer guarantees around match performance (compared to saying "if you only use node matchers you will get reasonable performance").

Reply:
I'll start a doc to put together some examples using both the currently proposed API and an alternative API that offers no distinction between node and leaf matchers.

Reply:
https://docs.google.com/document/d/1PRKQ8WNgNM1IVQt9Na9X31xilwIHyZcBbJ8vhdgVUfI/edit?usp=sharing has a few examples, lmk if you think there are more examples you'd like to see. 

I don't have the API specced out for joining the leaf and node protos into one, but I'm thinking we'd just have all the different matchers in a big oneof and have all the matchers be able to either return an action or another subtree.

Reply:
Thanks a left a few comments on the doc. It's helpful, but I find the examples to be confusing to read. Can you add more description and/or YAML comments and we can discuss there?